### ISSUE TITLE
Pipelining zero bubble and activation checkpointing bug

### ISSUE THREAD

**Issue by H-Huang** [pytorch/pytorch#136766]
There is a weird interaction with zero bubble and activation checkpointing which causes training to OOM. I don't believe this is a memory leak because for small models i dont see memory increasing across iterations, but I think this issue happens activation checkpointing probably has some incompatibility with the `stage_backward_input` implementation where we are going through the autograd graph in order to create a graph of the intermediates. I'm not sure yet, looking more into it.

Here are some memory traces

Interleaved 1F1B (over 10 steps):
<img width="1509" alt="image" src="https://github.com/user-attachments/assets/45b1f8a9-93de-4aaf-ac9f-d3bb25b185b5">

Zero bubble (eventually OOMs before 1 step is completed):
<img width="1714" alt="image" src="https://github.com/user-attachments/assets/cbf123b4-4369-49c0-a5d0-19109ec48fc9">

cc @XilunWu @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

**Comment by ufotalent** [comment]
Hi I'm one of the authors of zero bubble paper. We're really appreciated and excited for your effort to integrate zero bubble into pytorch!
Is it possible that some activation memory is recomputed in B pass (input grad) but freed after W (weight) passes? This might increase the memory. But however as long as we're not delaying all W passes to the end of iteration, there should still be a stable pattern of memory (instead of constantly increasing) though.

**Comment by wconstab** [comment]
I'll let @H-Huang comment but I believe he found the root causes and fixed, so we can close this issue. But generally, it would be great to get your feedback on this feature overall. We are also working on implementing a ZB-V style schedule. @haocizhang has a PR for that.

**Comment by H-Huang** [comment]
Hi @ufotalent, thanks for commenting and I enjoyed reading your paper 🙂. As @wconstab mentioned we fixed this issue in https://github.com/pytorch/pytorch/pull/138119 and it was a bug in our implementation. 

We would appreciate any feedback on our implementation of your paper. If you would like to take a look here is the stack of PRs (https://github.com/pytorch/pytorch/pull/132691) as we added a novel implementation of autograd splitting which allows you to compute "backward with respect to input" and "backward with respect to weight" separately so this works for all models. 

I will close this issue for now, but please create a new issue or comment to let us know if you have any questions or additional inputs, thanks!

**Comment by ufotalent** [comment]
Hi @H-Huang that PR of native B-W split is wonderful! I personally tried but didn't figure out how to have a generic implementation of splitting input grad and weight grad. Apparently you guys are more familiar with the autograd logic, lol.

I do have a question as I see there's a `retain_graph=True` on the input grad pass. Are the intermediate gradients that are NOT needed in the weight grad pass automatically released after the input grad pass? Or do we still need to wait until the weight grad pass for them to release?

Thanks

**Comment by H-Huang** [comment]
Hi @ufotalent

> I do have a question as I see there's a retain_graph=True on the input grad pass. Are the intermediate gradients that are NOT needed in the weight grad pass automatically released after the input grad pass?

Yes, the intermediate gradients that are NOT needed are still being released. `retain_graph=True` ensures that the autograd nodes are kept alive but these have weakrefs to the gradients, so if there are no references to them the unused intermediates can still be released. The useful intermediates that are kept alive belong in the `param_groups` data structure that gets passed into weight grad pass. Once that is finished then those are released as well!
